source/Color.cpp

```diff
@@ -130,3 +136,66 @@ Color Color::Multiply(float scalar, const Color &base)
 			scalar * base.color[2],
 			scalar * base.color[3]);
 }
+
+
+
+// Apply color blindness filters to this color.
+Color Color::Filter(Color c)
+{
+	Preferences::ColorFilter filter = Preferences::GetColorFilterMode();
+	// No color filters are enabled, so there's no need to adjust anything.
+	if(filter == Preferences::ColorFilter::NORMAL)
+		return c;
+
+	// Color blindness accessibility filters are enabled.
+	// LMS Daltonization is used to make the colors more distinct for
+	// color-blind players.
+
+	// Convert the colors from RGB to LMS, a color space that represents the
+	// light received by the cones in the human eye better than RGB.
+	float l = (17.8824 * c.color[0]) + (43.5161 * c.color[1]) + (4.11935 * c.color[2]);
+	float m = (3.45565 * c.color[0]) + (27.1554 * c.color[1]) + (3.86714 * c.color[2]);
+	float s = (0.0299566 * c.color[0]) + (0.184309 * c.color[1]) + (1.46709 * c.color[2]);
+
+	// Simulate color blidness.
+	switch(filter)
+	{
+		case Preferences::ColorFilter::PROTANOPIA:
+			l = (2.02344 * m) + (-2.52581 * s);
+			break;
+		case Preferences::ColorFilter::DEUTERANOPIA:
+			m = (0.494207 * l) + (1.24827 * s);
+			break;
+		case Preferences::ColorFilter::TRITANOPIA:
+			s = (-0.395913 * l) + (0.801109 * m);
+			break;
+		case Preferences::ColorFilter::NORMAL:
+			break;
+	}
+
+	// Convert the LMS colors back to RGB.
+	float r = (0.0809444479 * l) + (-0.130504409 * m) + (0.116721066 * s);
+	float g = (-0.0102485335 * l) + (0.0540193266 * m) + (-0.113614708 * s);
+	float b = (-0.000365296938 * l) + (-0.00412161469 * m) + (0.693511405 * s);
+
+	// Compensate for invisible colors.
+	r = c.color[0];
```

Comment:
I don't think this is correct.
This appears to just discard the value that was just placed here?
From what I've read on this algorithm, the end result may not actually change the red component of the colour, we do still need the red component we got back from the lms conversion for calculating the final green and blue values.
Currently, the next two lines do: `c.color[0] - r`, but `r` was just assigned `c.color[0]`, so the difference is always 0, and it doesn't look to me like that's actually what's supposed to happen.

Reply:
Whoops, you're right...
Although, putting it after decreased the contrast in some cases, so I ended up decreasing the coefficient on `r` to 0.5 instead of 0.7, which seems to handle things better (for the compensation for `g` and `b`).

Reply:
Actually, it seems to work better without taking the `r` value into compensation at all, so I'm just going to switch things back but remove the useless code.

Reply:
Define "work better." How are you determining if your implementation works for people with color blindness? Why make changes to an algorithm that you said is pretty standard?